scalableminds/webknossos
# Transactional Batch Saving

### URL of deployed dev instance (used for testing):
- https://savebatchtransactions.webknossos.xyz

### Steps to test:
- drag large NML into wK
- save, leave page before it reaches 100%
- reload page, should get old or new tracing (not something in between)
- kill backend during saving
- retry should either continue seamlessly or show conflict and the upcoming reload should give the old (consistent) tracing – (which of these two depends on the timing of the backend kill)
- the only way to still get inconsistent tracings, i think, is to kill the backend during the committing phase when it received the final request of a transaction
- when having a tracing open in two tabs, one (A) saving a large NML, and one (B) saving a small change after A has already successfully sent some requests, the small change (B) should still win (commits first) and the batched large saving process from A should fail when committing (refresh should then yield the small change from B)
- normal saving should also still work (both skeleton + volume)
- normal saving combined with a large transaction should also work

### Issues:
- fixes #3694 

### TODO
- [x] sync batch cache to disk (redis?)
- [x] ignore duplicate requests properly again (see #3767 )
- [x] clean up code, extract functions?
- [x] remove redundant request id
- [x] add config for redis address
- [x] in health check, try redis
- [x] redis error handling?
- [x] update readme (redis is now required)
- [x] discuss if progress percentage should be adapted (committing requests are significantly slower on the server) (→for now, let’s leave it as is)
- [x] fix front-end tests
- [x] include redis in docker files
- [x] adapt infrastructure setup to include redis
- [x] test health on a dev instance
- [x] test all the scenarios
- [x] after testing, remove (some of) the debug output


------
- [x] Ready for review (backend)
